zmon_aws_agent/elastigroup.py

```diff
@@ -0,0 +1,228 @@
+        return False
+
+
+@trace(tags={'aws': 'elastigroup'}, pass_span=True)
+def get_elastigroup_entities(region, acc, **kwargs):
+    groups = []
+    current_span = extract_span_from_kwargs(**kwargs)
+    current_span.set_tag("aws_region", region)
+    current_span.set_tag("account_id", acc)
+    try:
+        cf = boto3.client('cloudformation', region_name=region)
+
+        stack_names = get_all_stack_names(cf)
+        for stack_name in stack_names:
+            elastigroups = get_elastigroup_resources(cf, stack_name)
+            for eg_data in elastigroups:
+                eg_details = get_elastigroup(eg_data, **kwargs)
+                eg_name = eg_details.get('name', eg_details.get('id', 'unknown-elastigroup'))
+                capacity = eg_details.get('capacity', {})
+                strategy = eg_details.get('strategy', {})
+                compute = eg_details.get('compute', {})
+                eg = {
+                    'id': entity_id('elastigroup-{}[{}:{}]'.format(eg_name, acc, region)),
+                    'type': 'elastigroup',
+                    'infrastructure_account': acc,
+                    'region': region,
+                    'created_by': 'agent',
+                    'name': eg_name,
+                    'availability_zones': [az.get('name', 'unknown-az') for az in
+                                           eg_details.get('compute', {}).get('availability_zones', [])],
+                    'desired_capacity': capacity.get('target', 1),
+                    'max_size': capacity.get('maximum', 1),
+                    'min_size': capacity.get('minimum', 1),
+                    'risk': strategy.get('risk', 100),
+                    'orientation': strategy.get('availability_vs_cost', 'balanced'),
+                    'instance_types': compute.get('instance_types', None),
+                    'created_time': eg_details.get('created_at', None),
+                }
+
+                for tag in compute.get('launch_specification', {}).get('tags', []):
+                    key = inflection.underscore(tag.get('tag_key', None))
+                    val = tag.get('tag_value', None)
+                    if key and val and key not in eg:
+                        eg[key] = val
+                add_traffic_tags_to_entity(eg)
+
+                eg['instances'] = []
+                instances = get_elastigroup_instances(eg_data)
+                for instance in instances:
+                    eg['instances'].append({'aws_id': instance.get('instance_id', 'missing-instance-id'),
+                                            'ip': instance.get('private_ip', 'missing-private-ip')})
+
+                groups.append(eg)
+    except Exception as e:
+        if isinstance(e, ClientError) and e.response['Error']['Code'] == 'AccessDenied':
+            msg = 'Access to required AWS API denied. Skipping Elastigroup discovery.'
+            logger.warning(msg)
+            current_span.log_kv({'message': msg})
```

Comment:
May be `current_span.set_tag('permission_error', True)` would be helpful in general. Can be combined with `aws: elastigroup` existing tag.
Also we can have the next part common regardless of the exception type:
```python
            current_span.set_tag('error', True)
            current_span.log_kv({'exception': traceback.format_exc()})
            if isinstance(e, ClientError) and e.response['Error']['Code'] == 'AccessDenied':
           ...
           else:
```